File: Max_Flow/cycle_Detection.java
package Max_Flow;

import java.util.ArrayList;

public class cycle_Detection {
    public static ArrayList<Edge> graph[];
    public static boolean vis[];
    public static boolean res[];
    
    public static boolean isCycle(int curr){
        vis[curr]=true;
        res[curr]=true;
        for(int a=0;a<graph[curr].size();a++){
            Edge e=graph[curr].get(a);
            if(res[e.dest]){
                return true;
            }
            else if(!vis[e.dest]){
                if(isCycle(e.dest)){
                    return true;
                }
            }
        }
        res[curr]=false;
        return false;
    }

    public static int v;
    static class Edge{
        int src;
        int dest;
        int weight;
        public Edge(int src, int dest, int weight){
            this.src=src;
            this.dest=dest;
            this.weight=weight;
        }
    }
    
    public static void createGraph(ArrayList<Edge> graph[]){
        for(int a=0;a<v;a++){
            graph[a]=new ArrayList<>();
        }
        
        graph[0].add(new Edge(0,2,1));
        graph[1].add(new Edge(1,0,1));
        graph[2].add(new Edge(2,3,1));
        graph[3].add(new Edge(3,0,1));
    }
    
    public static void main(String[] CSECU){
        v=4;
        graph = new ArrayList[v];
        createGraph(graph);
        /*
         1---0---3
             /  /
             / /
             //
             2
         */
        vis=new boolean[v];
        res=new boolean[v];
        for(int a=0;a<v;a++){
            if(!vis[a]){
                boolean ans=isCycle(a);
                if(ans){
                    System.out.println("Cycle");
                    return;
                }
            }
        }
        System.out.println("No Cycle");
    }
}

File: Max_Flow/Kotlin/CycleDetection.kt
package Max_Flow

import java.util.*

class CycleDetection {
    data class Edge(val src: Int, val dest: Int, val weight: Int)

    companion object {
        private lateinit var graph: Array<ArrayList<Edge>>
        private lateinit var vis: BooleanArray
        private lateinit var res: BooleanArray
        private var v: Int = 0

        private fun isCycle(curr: Int): Boolean {
            vis[curr] = true
            res[curr] = true
            for (e in graph[curr]) {
                if (res[e.dest]) {
                    return true
                } else if (!vis[e.dest]) {
                    if (isCycle(e.dest)) {
                        return true
                    }
                }
            }
            res[curr] = false
            return false
        }

        private fun createGraph(graph: Array<ArrayList<Edge>>) {
            for (a in 0 until v) {
                graph[a] = ArrayList()
            }

            graph[0].add(Edge(0, 2, 1))
            graph[1].add(Edge(1, 0, 1))
            graph[2].add(Edge(2, 3, 1))
            graph[3].add(Edge(3, 0, 1))
        }

        @JvmStatic
        fun main(args: Array<String>) {
            v = 4
            graph = Array(v) { ArrayList() }
            createGraph(graph)
            /*
             1---0---3
                 /  /
                 / /
                 //
                 2
             */
            vis = BooleanArray(v)
            res = BooleanArray(v)
            for (a in 0 until v) {
                if (!vis[a]) {
                    val ans = isCycle(a)
                    if (ans) {
                        println("Cycle")
                        return
                    }
                }
            }
            println("No Cycle")
        }
    }
} 